Added beta specs docs (#1977)
holochain/holochain: docs/specs/src/hwp_2_overview.md

Holochain Design Overview: A Game Play Metaphor
===============================================

It may help to understand the design of Holochain through a well known
pattern of agentic collaboration: playing games.

Playing Games
-------------

People define the rules of a *Game* they want to play together. As

> *Players* join the *Game*, they alter their state by appending records

Author: I've updated this to:
 `As
*Players* join the *Game* and start playing, they append new *Records* of 
their play to their own *Action* history. Then they update 
the *Game Board* by sharing the *Records* of their *Actions*[^chess] with
other players.`

> *Players* join the *Game*, they alter their state by appending records
> to their *Action* history. Then they update the *Game Board* by sharing
> the *Records* of their *Actions*[^4].

Reviewer: Perhaps this is a bit easier and still accurate:

```suggestion
*Players* join the *Game*, they alter their state by appending *Actions*
to their history. Then they update the *Game Board* by sharing
the *Records* of their *Actions*[^4].
```

The first requirement to create social coherence is ensuring that people
are playing the same game, therefore the very first record in every
Agent's history is the rules of the game by which they agree to play.
Obviously, Players are not in the same game or able to use the same Game
Board if they don't start with the same ruleset. These rules are the
actual computer code that is executed in running the Game, making moves
and validating the Actions of all Players.

System Description
------------------

> We can describe the system as Agents, who play Games together by taking

Reviewer: I find the comma more readable, personally

Actions, the Records of which are held in a distributed Ledger that is
built by sharing these Records over a Network with other Agents. We
capitalize terms that comprise the ontological units of the system, and
which are formally described in the later sections.

### Agents

Agents have these properties:

1.  Agents are the only source of Actions in the system, thus Agents are
  the source of agency. All such Actions are uniquely identifiable

>   as to which Agent took them, i.e. all Actions are
>   cryptographically signed (see Actions below).

Reviewer: ```suggestion
  as to which Agent took them, i.e. all Actions are signed 
  by public-key cryptography (see Actions below).
```

  Action which is also recorded. Note: both types of data,
  content of the play (Entry) and the meta-data about the play
  (Action) when taken together are called a Record.

    b.  A deterministic description of the validity of the contents of
  an Entry along with the validity of an Agent taking any
  Action, i.e. some Actions may to be taken in some contexts but
  not in others, as in making a move out-of-turn, or by an Agent
  not allowed to play a certain role.

    c.  A deterministic description of how Records may be related or
  linked to other Records.

2.  Along with the Integrity specification, a Game also consists of a
  Coordination specification. This specification contains
  instruction sets that wrap Actions into function call units and
  thus serve as an API to the Game. For example, for a blogging
  "Game" one such function call might be "create\_post" which takes
  a number of Actions that atomically create a number of Records to
  the Agent's Source Chain which include an Entry for the post as
  well as links relating the post to other Entries (see below for
  definitions of Actions and Source Chain).

> 3.  Each instance of the Game is played on it's own Game Board which

Reviewer: ```suggestion
3.  Each instance of the Game is played on its own Game Board which
```

  exists as a unique and independent network of Agents playing that
  Game, i.e. Games cannot interact with each other directly as all
  action in the system is only taken by Agents. Note that Games can

>   be composed together, but only by Agents playing across multiple

Reviewer: I don't feel like this needs any extra explanation, but if there were more, maybe it could be stated that the "new game" which is created by the composition of two existing games only exists for the agents who are playing both games. Like, the game of making puns in German by inserting English words that sound like German words can only be played by people who speak both languages.

>   be composed together, but only by Agents playing across multiple

Author: What this is intended to point to set-up for is UIs that are assemblages of multiple DNAs.  I think I'll add a couple more sentences to make this clearer.

  games. This at first may seem like a weakness, but it's part of a
  key design decision that contributes to the system's overall
  design goals of evolvability.

In keeping with the metaphor of Game, we also refer to the Integrity
specification as the Validation rules of the Game.

We also refer to both the Integrity and Coordination specifications of a
Game as its DNA because this evokes the pattern of all the "cells" in
the social "body" as being built out of the same instruction set, thus
being the ground of self for that social body.

### Actions (and Entries and Records)

Actions have these properties:

1.  An Action has cryptographic provenance in that it is signed by the
  Agent that took the Action.

2.  Actions are Recorded in a monotonically temporally increasing
  hash-chain by the Agent that takes the Action. We refer to this as
  a hash-chain because each Action includes in it the hash of the
  previous Action, thus creating an untamperable Action history.

3.  Actions are addressed by the hash of the Action.

4.  There are a number of Action types:

```{=html}
<!-- -->

> ```

Reviewer: does this need to be here for formatting or something?

Author: Yes, without it the next markdown list doesn't get appropriately indented.

1.  CreateEntry: An Action for adding new Game-specific content. We call
  such content an Entry. The Entry may be declared as public, and
  will thus be published by the Agent to the network, or declared as
  private, where publishing is limited to just the Action data and
  not the content. Entries are addressed by their hash, and thus for
  CreateEntry Actions, this hash is included in it. Thus sometimes
  the Action is considered to be "meta-data" where the Entry is

>   considered "data"[^5] .

Reviewer: I might also add some corollaries about Entries:

- They are independently addressable by design, so that Actions can be referenced independently of their Entries
- There may be many Actions that "create" a single Entry, so in general there is no way to unambiguous way to assign an Action to a particular Entry

>   considered "data"[^5] .

Author: added

2.  UpdateEntry: An Action which adds new Game-specific content onto the
  chain that is intended to update previous content.

3.  MarkEntryAsDeleted: An Action which indicates a previous entry
  should be considered deleted.

> 4.  CreateLink: An Action that links one hash to another

Reviewer: ```suggestion
4.  CreateLink: An Action that unidirectionally links one hash to another
```

5.  DeleteLink: An Action that indicates a previous link action should
  be considered deleted.

> 6.  A Record is just a name for both an Action and its Entry taken
>   together. As an implementation detail, note that for actions other
>   than CreateEntry and UpdateEntry we don't need to address the
>   content of the Action separately and so we simply retrieve all the
>   data we need from the recorded Action.

Reviewer: ```suggestion
6.  A Record is just a name for both an Action and, when applicable, 
  its Entry, taken together. As an implementation detail, note that for 
  actions other than CreateEntry and UpdateEntry we don't need to 
  address the content of the Action separately, in which case the Record 
  contains no Entry and we simply retrieve all the we simply retrieve all the
  data we need from the recorded Action.
```

a DHT to include linking hashes to other hashes, thus creating a
Graphing DHT.

As a distributed database the DHT may be understood as a transform of
Agent's Source-Chain state into a form that makes that data retrievable
by all the other Agents for different purposes. These purposes include:

1.  Retrieval of Agent's Actions and created Entries

2.  Confirmation of "good behavior" by retrieving an Agent's activity
  history which is used to verify that agents haven't forked their
  chains

3.  Retrieval of link information

4.  Retrieval of validation receipts

To achieve this end, we take advantage of the fact that an Agent's
public key (which serves as its address) is in the same numeric space as
the hashes of the data that we want to store and retrieve. Using this
property, we can create a mapping between Agents responsible for holding
portions of the overall data by using a nearness algorithm between the
Agent's public key and the hash of the data to be stored. Agents that
are "close" to a given piece of data are responsible to store it and are
said to comprise a Neighborhood for that data. Hashing creates an

> essentially random distribution of which data will be stored with
> Agents. The degree of redundancy of how many Agents should store copies

Reviewer: ```suggestion
essentially random distribution of which data will be stored with which
Agents. The degree of redundancy of how many Agents should store copies
```

of data is a per-Game parameter.

Agents periodically gossip with other Agents in their Neighborhood about
the published data they've received, validating and updating their
Records accordingly. This gossip ensures that eventually all Agents
querying a Neighborhood for information will receive the same
information. Furthermore it creates a social space for detecting bad
actors. Because all gossiped data can intrinsically be validated, any

> Agents who cheat, including by changing their (or other's) history's,

Reviewer 2: ```suggestion
Agents who cheat, including by changing their (or other's) histories,
```
